### ISSUE TITLE
Undefined Reference to LoadLibrary

### ISSUE THREAD

**Issue by Zulsorai** [lunarmodules/luasocket#296]
I've been trying to install LuaSocket via LuaRocks and I keep getting a compile failure with *socket/core.dll*.  
I'm running Windows with MinGW and Lua 5.3.
<details>
  <summary>Log</summary>

```
> luarocks make

mingw32-gcc -O2 -c -o src/luasocket.o -IC:/Dev/LUA/5.3.5/include src/luasocket.c -DLUASOCKET_DEBUG -DLUASOCKET_INET_PTON -DWINVER=0x0501 -Ic:/windows/system32/include
mingw32-gcc -O2 -c -o src/timeout.o -IC:/Dev/LUA/5.3.5/include src/timeout.c -DLUASOCKET_DEBUG -DLUASOCKET_INET_PTON -DWINVER=0x0501 -Ic:/windows/system32/include
mingw32-gcc -O2 -c -o src/buffer.o -IC:/Dev/LUA/5.3.5/include src/buffer.c -DLUASOCKET_DEBUG -DLUASOCKET_INET_PTON -DWINVER=0x0501 -Ic:/windows/system32/include
mingw32-gcc -O2 -c -o src/io.o -IC:/Dev/LUA/5.3.5/include src/io.c -DLUASOCKET_DEBUG -DLUASOCKET_INET_PTON -DWINVER=0x0501 -Ic:/windows/system32/include
mingw32-gcc -O2 -c -o src/auxiliar.o -IC:/Dev/LUA/5.3.5/include src/auxiliar.c -DLUASOCKET_DEBUG -DLUASOCKET_INET_PTON -DWINVER=0x0501 -Ic:/windows/system32/include
mingw32-gcc -O2 -c -o src/options.o -IC:/Dev/LUA/5.3.5/include src/options.c -DLUASOCKET_DEBUG -DLUASOCKET_INET_PTON -DWINVER=0x0501 -Ic:/windows/system32/include
mingw32-gcc -O2 -c -o src/inet.o -IC:/Dev/LUA/5.3.5/include src/inet.c -DLUASOCKET_DEBUG -DLUASOCKET_INET_PTON -DWINVER=0x0501 -Ic:/windows/system32/include
mingw32-gcc -O2 -c -o src/except.o -IC:/Dev/LUA/5.3.5/include src/except.c -DLUASOCKET_DEBUG -DLUASOCKET_INET_PTON -DWINVER=0x0501 -Ic:/windows/system32/include
mingw32-gcc -O2 -c -o src/select.o -IC:/Dev/LUA/5.3.5/include src/select.c -DLUASOCKET_DEBUG -DLUASOCKET_INET_PTON -DWINVER=0x0501 -Ic:/windows/system32/include
mingw32-gcc -O2 -c -o src/tcp.o -IC:/Dev/LUA/5.3.5/include src/tcp.c -DLUASOCKET_DEBUG -DLUASOCKET_INET_PTON -DWINVER=0x0501 -Ic:/windows/system32/include
mingw32-gcc -O2 -c -o src/udp.o -IC:/Dev/LUA/5.3.5/include src/udp.c -DLUASOCKET_DEBUG -DLUASOCKET_INET_PTON -DWINVER=0x0501 -Ic:/windows/system32/include
mingw32-gcc -O2 -c -o src/compat.o -IC:/Dev/LUA/5.3.5/include src/compat.c -DLUASOCKET_DEBUG -DLUASOCKET_INET_PTON -DWINVER=0x0501 -Ic:/windows/system32/include
mingw32-gcc -O2 -c -o src/wsocket.o -IC:/Dev/LUA/5.3.5/include src/wsocket.c -DLUASOCKET_DEBUG -DLUASOCKET_INET_PTON -DWINVER=0x0501 -Ic:/windows/system32/include
mingw32-gcc -shared -o socket/core.dll src/luasocket.o src/timeout.o src/buffer.o src/io.o src/auxiliar.o src/options.o src/inet.o src/except.o src/select.o src/tcp.o src/udp.o src/compat.o src/wsocket.o -Lc:/windows/system32 -lws2_32 C:/Dev/LUA/5.3.5/lib/lua53.dll -lmc:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _GetModuleHandleA@4 by linking to _GetModuleHandleA
Use --enable-stdcall-fixup to disable these warnings
Use --disable-stdcall-fixup to disable these fixups
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _GetProcAddress@8 by linking to _GetProcAddress
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _FreeLibrary@4 by linking to _FreeLibrary
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _Sleep@4 by linking to _Sleep
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _GetSystemTimeAsFileTime@4 by linking to _GetSystemTimeAsFileTime
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _setsockopt@20 by linking to _setsockopt
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _htonl@4 by linking to _htonl
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _getsockopt@20 by linking to _getsockopt
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _inet_ntoa@4 by linking to _inet_ntoa
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _getaddrinfo@16 by linking to _getaddrinfo
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _getnameinfo@28 by linking to _getnameinfo
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _freeaddrinfo@4 by linking to _freeaddrinfo
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _gethostname@8 by linking to _gethostname
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _getpeername@12 by linking to _getpeername
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _FormatMessageA@28 by linking to _FormatMessageA
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _getsockname@12 by linking to _getsockname
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving ___WSAFDIsSet@8 by linking to ___WSAFDIsSet
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _WSAStartup@8 by linking to _WSAStartup
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _WSACleanup@0 by linking to _WSACleanup
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _select@20 by linking to _select
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _WSAGetLastError@0 by linking to _WSAGetLastError
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _ioctlsocket@12 by linking to _ioctlsocket
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _closesocket@4 by linking to _closesocket
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _shutdown@8 by linking to _shutdown
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _socket@12 by linking to _socket
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _connect@12 by linking to _connect
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _bind@12 by linking to _bind
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _listen@8 by linking to _listen
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _accept@12 by linking to _accept
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _send@16 by linking to _send
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _sendto@24 by linking to _sendto
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _recv@16 by linking to _recv
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _recvfrom@24 by linking to _recvfrom
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _gethostbyaddr@12 by linking to _gethostbyaddr
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _gethostbyname@4 by linking to _gethostbyname
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _EnterCriticalSection@4 by linking to _EnterCriticalSection
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _TlsGetValue@4 by linking to _TlsGetValue
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _GetLastError@0 by linking to _GetLastError
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _LeaveCriticalSection@4 by linking to _LeaveCriticalSection
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _DeleteCriticalSection@4 by linking to _DeleteCriticalSection
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _InitializeCriticalSection@4 by linking to _InitializeCriticalSection
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _VirtualQuery@12 by linking to _VirtualQuery
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: warning: resolving _VirtualProtect@16 by linking to _VirtualProtect
c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../../mingw32/bin/ld.exe: c:/dev/mingw/bin/../lib/gcc/mingw32/8.2.0/crtbegin.o:cygming-crtbegin.c:(.text+0x25): undefined reference to `LoadLibraryA@4'
collect2.exe: error: ld returned 1 exit status

Error: Build error: Failed compiling module socket/core.dll
```
</details>

I've been searching around for potential fixes but haven't had any luck. The closest I managed to find was #124 but it wasn't any help. 
Any advice would be appreciated.

**Comment by SunMaojie** [comment]
I had the same problem， faillue with  socket/core.dll
windows10 with MinGW and Lua5.3.5

**Comment by alerque** [comment]
This looks like a combination of some things that have since gotten fixed in our Lua 5.3 (and now 5.4) builds and some miss-matched headers on the original system being used.

If that estimation is wrong feel free to comment with some current details for review and I'll re-open, but for now I think there is nothing here to act on.
